tests/bot/cogs/moderation/test_utils.py

```diff
@@ -0,0 +1,411 @@
+
+                send_private_embed_mock.assert_awaited_once_with(args[0], embed)
+
+    @patch("bot.cogs.moderation.utils.log")
+    async def test_post_user(self, log_mock):
+        """Should POST a new user and return the response if successful or otherwise send an error message."""
+        user = MockUser(discriminator=5678, id=1234, name="Test user")
+        some_mock = MagicMock(discriminator=3333)
+        test_cases = [
+            {
+                "user": user,
+                "post_result": "bar",
+                "raise_error": None,
+                "payload": {
+                    "discriminator": 5678,
+                    "id": self.user.id,
+                    "in_guild": False,
+                    "name": "Test user",
+                    "roles": []
+                }
+            },
+            {
+                "user": self.member,
+                "post_result": "foo",
+                "raise_error": ResponseCodeError(MagicMock(status=400), "foo"),
+                "payload": {
+                    "discriminator": 0,
+                    "id": self.member.id,
+                    "in_guild": False,
+                    "name": "Name unknown",
+                    "roles": []
+                }
+            },
+            {
+                "user": some_mock,
+                "post_result": "bar",
+                "raise_error": None,
+                "payload": {
+                    "discriminator": some_mock.discriminator,
+                    "id": some_mock.id,
+                    "in_guild": False,
+                    "name": some_mock.name,
+                    "roles": []
+                }
+            }
+        ]
+
+        for case in test_cases:
+            test_user = case["user"]
+            expected = case["post_result"]
+            error = case["raise_error"]
+            payload = case["payload"]
+
+            with self.subTest(user=test_user, result=expected, error=error, payload=payload):
+                log_mock.reset_mock()
+                self.bot.api_client.post.reset_mock(side_effect=True)
+                self.ctx.bot.api_client.post.return_value = expected
+
+                self.ctx.bot.api_client.post.side_effect = error
+
+                result = await utils.post_user(self.ctx, test_user)
+
+                if error:
+                    self.assertIsNone(result)
+                else:
+                    self.assertEqual(result, expected)
+
+                if not error:
+                    self.bot.api_client.post.assert_awaited_once_with("bot/users", json=payload)
+                else:
+                    self.ctx.send.assert_awaited_once()
+                    self.assertTrue(str(error.status) in self.ctx.send.call_args[0][0])
+
+                if isinstance(test_user, MagicMock):
+                    log_mock.debug.assert_called_once()
+                else:
+                    log_mock.debug.assert_not_called()
```

Comment:
I don't like the amount of branching here too much.
```python
if error:
  ...
else:
  ...
if not error:
  ...
else:
  ...
```
Can these be joined together?

You marked @MarkKoz's comment above as resolved, but is it? Why are you only checking if `post` was awaited if it won't raise?

Reply:
Additionally I'm finding this fairly awkward:
```python
if isinstance(test_user, MagicMock):
```
I don't think it's necessary to test whether a `debug` log was made (if that doesn't indicate that some other action was also performed). But it can stay. I'm just wondering whether we could come up with a nicer way of branching here.

Reply:
Removed `debug` call assertion. Joined these if statements together. About only asserting `post`, what else this does? I'm also asserting returned result, but this calls nothing else what I could assert.